pkgs/build-support/setup-hooks/auto-patchelf.py

```diff
@@ -0,0 +1,316 @@
+                    continue
+
+                try:
+                    with openELF(os.path.join(root, file)) as elf:
+                        osabi = getBinOsabi(elf)
+                        arch = getBinArch(elf)
+
+                        if rpath := getRpathFromElfBinary(elf):
+                            libDirs += rpath
+                        #print("Caching", file, arch, lib, osabi)
+
+                        sonameCache[(file, arch)].append( (root, osabi) )
+                        used = True
+                except ELFError:
+                    pass
+            if used:
+                print("found .so libs in", root)
+
+            if not recurse:
+                break;
+
+
+def findDependency(soname, arch, osabi):
+    for lib, libabi in sonameCache[(soname, arch)]:
+        if areBinOsabisCompatible(osabi, libabi):
+            return lib
+    return None
+
+def autoPatchelfFile(elf, runtimeDeps):
+    fileArch = getBinArch(elf)
+    fileOsabi = getBinOsabi(elf)
+    rpath = []
+
+    if interpreterArch != fileArch:
+        # Our target architecture is different than this file's architecture, so skip it.
+        print("skipping {} because its architecture ({}) differs from target ({})"
+            .format(elf.filename, fileArch, interpreterArch))
+        return
+    elif not areBinOsabisCompatible(interpreterOsabi, fileOsabi):
+        print("skipping {} because its OS ABI ({}) is not compatible with target ({})"
+            .format(elf.filename, fileOsabi, interpreterOsabi))
+        return
+
+    if isExecutable(elf):
+        print("setting interpreter of", elf.filename)
+        subprocess.run(["patchelf", "--set-interpreter", interpreterPath, elf.filename], check = True)
+        rpath += ["{}/lib".format(path) for path in runtimeDeps]
+
+    print("searching for dependencies of", elf.filename)
+    try:
+        missing = getDepsFromElfBinary(elf)
+    except ELFError:
+        return
+
+    # This ensures that we get the output of all missing dependencies instead
+    # of failing at the first one, because it's more useful when working on a
+    # new package where you don't yet know its dependencies.
+
+    for dep in missing:
+        if dep.startswith("/"):
+            # This is an absolute path. If it exists, just use it. Otherwise,
+            # we probably want this to produce an error when checked (because
+            # just updating the rpath won't satisfy it).
+            if os.path.isfile(dep):
+                continue
+        elif os.path.isfile(os.path.join(libcLib, dep)):
+            # This library exists in libc, and will be correctly resolved by
+            # the linker.
+            continue
+
+        if foundDependency := findDependency(dep, fileArch, fileOsabi):
+            rpath.append(foundDependency)
+            print("    {} -> found: {}".format(dep, foundDependency))
+        else:
+            print("    {} -> not found!".format(dep))
+            yield dep
```

Comment:
Not personally a fan of this, since it does make the functions behaviour less obvious. Without knowing/looking at its implementation, what would you say `autoPatchelfFile` would yield or return?

Reply:
Good point. And at the same time making that a global variable is not so nice either. Maybe it makes sense here.

Reply:
I'd probably return a [`dataclass`](https://docs.python.org/3/library/dataclasses.html) or [`namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple), eg. like this:

```python
@dataclass
class Dependency:
    path: Path
    found: bool

def auto_patchelf_file(path: Path, deps: List[Dependency]) -> List[Dependency]:
    ...
```

Reply:
I used a global in the end. emphasis is on testing ATM. I also struggle with type annotations. What is the right way to check/enforce them ?

Reply:
done :-). My global was not really global, and usign global (the keyword) is ugly and error prone.